TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeMedicalHistory(with session: Session, modal: Flow) -> UIViewController
+        func makeMedicalHistory(modal: Flow) -> UIViewController
+    }
+}
+```
+
+This would remove the intricated structure we have with builders and its children.
+
+A side effect of this approach, because we are dealing with functions (versus structs) we are able to remove local state. An example of such case is `enableClinicalDownloads` at `ClinicalRecordsChildBuilders`. You can see the code smell here:
+
+```swift
+builders.appointmentList(with: session,
+                         enableClinicalDownloads: builders.enableClinicalDownloads,
+                         addressSwitches: builders.addressSwitches)
+```
+
+With the free functions approach, this would never be possible to do, since builders don't hold state:
+
+```swift
+extension Builder {
+    enum ClinicalRecords {
+        func makeAppointmentList(session: UserSession,
+                                 enableClinicalDownloads: Bool,
+                                 addressSwitches: AddressSwitchesProtocol,
+                                 primary: Flow,
+                                 secondary: Flow) -> UIViewController
+    }
+}
+```
+
+At the call site, we would replace this:
+
+```swift
+case let .appointmentList(session):
+    builders
+            .appointmentList(with: session,
+                             enableClinicalDownloads: builders.enableClinicalDownloads,
+                             addressSwitches: builders.addressSwitches)
+            .make(primary: navigation, secondary: modal)
+            |> navigation.present
+```
+
+With this:
+
+```swift
+case let .appointmentList(session):
+    Builder.ClinicalRecords
+            .makeAppointmentList(with: session,
+                                enableClinicalDownloads: enableClinicalDownloads,
+                                addressSwitches: addressSwitches,
+                                primary: navigation,
+                                secondary: modal)
+            |> navigation.present
+```
+
+## Impact on an existing codebase
+
+As Ilya pointed out it would make unit testing a bit more cumbersome. That's the case because you no longer inject builders. I would like to make the case for the following:
+
+1. We don't test `FlowControllers` with Unit tests anyway. This is not a reason to not do it, it's simply difficult to justify time doing so, when there are more important pieces to be tested and those tests would bring, in my opinion, little value. More value would be achieved by the next point (**2.**).
+2. We can test that a particular navigation/flow is done as expected independently of this proposal being approved or not. For that to happen we simply observe a ViewModel state and assert on it, since we derive routes from the state.
```

Comment:
That's what we already do in some view model tests, I think, but asserting routes is not the same as testing flows, it only tests view model part, but the bug can be (and we had such cases not that long time ago) in the flow controller.
Also considering proposal to use feedbacks for navigation we will not have routes any more.

Reply:
This proposal is independent of any other proposal.